module/signature/aggregation.go

```diff
@@ -3,21 +3,317 @@
+//  - ErrDuplicatedSigner if the signer has been already added
+//  - random error if the execution failed
+// The function does not return an error for any invalid signature.
+// If no error is returned, the bool represents the validity of the signature.
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) VerifyAndAdd(signer int, sig crypto.Signature) (bool, error) {
+	if signer >= s.n || signer < 0 {
+		return false, engine.NewInvalidInputErrorf("input index %d is invalid", signer)
+	}
+	_, duplicate := s.indexToSignature[signer]
+	if duplicate {
+		return false, newErrDuplicatedSigner("signer %d was already added", signer)
+	}
+	// signature is new
+	ok, err := s.publicKeys[signer].Verify(sig, s.message, s.hasher)
+	if ok {
+		s.indexToSignature[signer] = string(sig)
+	}
+	return ok, err
+}
+
+// TrustedAdd adds a signature to the internal state without verifying it.
+//
+// The Aggregate function makes a sanity check on the aggregated signature and only
+// outputs valid signatures. This would detect if TrustedAdd has added any invalid
+// signature.
+// The function errors:
+//  - engine.InvalidInputErrorf if the index input is invalid
+//  - ErrDuplicatedSigner if the signer has been already added
+//  - random error if the execution failed
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) TrustedAdd(signer int, sig crypto.Signature) error {
+	if signer >= s.n || signer < 0 {
+		return engine.NewInvalidInputErrorf("input index %d is invalid", signer)
+	}
+	_, duplicate := s.indexToSignature[signer]
+	if duplicate {
+		return newErrDuplicatedSigner("signer %d was already added", signer)
+	}
+	// signature is new
+	s.indexToSignature[signer] = string(sig)
+	return nil
+}
+
+// HasSignature checks if a signer has already provided a valid signature.
+//
+// The function errors:
+//  - engine.InvalidInputError if the index input is invalid
+//  - random error if the execution failed
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) HasSignature(signer int) (bool, error) {
+	if signer >= s.n || signer < 0 {
+		return false, engine.NewInvalidInputErrorf("input index %d is invalid", signer)
+	}
+	_, ok := s.indexToSignature[signer]
+	return ok, nil
+}
+
+// Aggregate aggregates the stored BLS signatures and returns the aggregated signature.
+//
+// Aggregate attempts to aggregate the internal signatures and returns the resulting signature.
+// The function performs a final verification and errors if any signature fails the desrialization
+// or if the aggregated signature is not valid. It also errors if no signatures were added.
+// required for the function safety since "TrustedAdd" allows adding invalid signatures.
+// The function is not thread-safe.
+//
+// TODO : When compacting the list of signers, update the return from []int
+// to a compact bit vector.
+func (s *SignatureAggregatorSameMessage) Aggregate() ([]int, crypto.Signature, error) {
```

Comment:
We should probably have a dedicated sentinel error for this case, not sure if we want to treat it as exception. If it fails it probably means that we have a bug in implementation, right?

Reply:
Currently, the function errors if:
- there is an unexpected low level error (memory for instance)
- the caller made `TrustedAdd` with an invalid signature - this would be a bug in Hotstuff, and it's not recoverable.

Reply:
So everything is an exception, sounds good to me.